File: sdk/typescript/src/pageindex.ts
/** Skills pageindex (markdown tree indexing and retrieval). */

import type { SkillLineContentRow } from "./types.js";

import {
  SkillsBuilderNative,
  buildPageIndexForFileNative,
  buildPageIndexOnlyNative,
  buildSkillsIndexNative,
  finalizeSkillDocumentJsonNative,
  getSkillContentRetrieveResultNative,
  getSkillDocumentNative,
  getSkillLineContentFromSpecNative,
  getSkillLineContentNative,
  getSkillStructureNative,
  loadMergedSkillDocumentJsonNative,
  loadSkillsIndexFromDirNative,
  loadSkillsIndexFromEntryNative,
  mdToTreeNative,
  pageIndexValidNative,
  parseSkillNodeIdsNative,
  tokenCountFromDecomposedFrontmatterNative,
  reconstructSkillMarkdownNative,
  repairSkillNodesNative,
  skillsIndexFromDecomposedDirNative,
  updateSkillDocumentSourcePathNative,
  writeReconstructedSkillNative,
  writeSkillsIndexNative,
  type ReconstructOptionsNapi,
} from "./native.js";

export interface PageIndexConfig {
  ifAddNodeId?: boolean;
  ifAddNodeText?: boolean;
}

/** CamelCase SDK config or snake_case partial dict (e.g. from app YAML). */
export type PageIndexConfigInput = PageIndexConfig | Record<string, unknown>;

export interface ReconstructOptions {
  keepAllHeaders?: boolean;
}

export interface SkillsIndexDict {
  documents: Record<string, unknown>;
  files: Record<string, string>;
}

export function defaultPageIndexConfig(): PageIndexConfig {
  return {
    ifAddNodeId: true,
    ifAddNodeText: false,
  };
}

/** Partial pageindex settings from app config; Rust merges unset keys with SDK defaults. */
export function pageIndexConfigFromMapping(
  mapping?: Record<string, unknown> | null,
): Record<string, unknown> | undefined {
  if (mapping == null) return undefined;
  if (isSnakeCasePageIndexDict(mapping)) return mapping;
  return pageIndexConfigToNative(pageIndexConfigFromPartial(mapping));
}

export function pageIndexConfigFromPartial(
  partial: Partial<PageIndexConfig> & Record<string, unknown>,
): PageIndexConfig {
  const cfg = defaultPageIndexConfig();
  if (partial.ifAddNodeId !== undefined) cfg.ifAddNodeId = partial.ifAddNodeId;
  if (partial.ifAddNodeText !== undefined)
    cfg.ifAddNodeText = partial.ifAddNodeText;
  return cfg;
}

export function pageIndexConfigToNative(
  config: PageIndexConfig,
): Record<string, unknown> {
  return {
    if_add_node_id: config.ifAddNodeId ?? true,
    if_add_node_text: config.ifAddNodeText ?? false,
  };
}

function isSnakeCasePageIndexDict(config: Record<string, unknown>): boolean {
  return "if_add_node_id" in config || "if_add_node_text" in config;
}

function resolveNativeConfig(
  config?: PageIndexConfigInput,
): Record<string, unknown> | undefined {
  if (config == null) return undefined;
  if (isSnakeCasePageIndexDict(config as Record<string, unknown>)) {
    return config as Record<string, unknown>;
  }
  return pageIndexConfigToNative(config as PageIndexConfig);
}

export function buildSkillsIndex(
  skillDirs: string[],
  config?: PageIndexConfigInput,
): SkillsIndexDict {
  return buildSkillsIndexNative(
    skillDirs,
    resolveNativeConfig(config),
  ) as SkillsIndexDict;
}

export function writeSkillsIndex(
  index: SkillsIndexDict,
  outputDir: string,
): void {
  writeSkillsIndexNative(index, outputDir);
}

export function loadSkillsIndexFromDir(catalogDir: string): SkillsIndexDict {
  return loadSkillsIndexFromDirNative(catalogDir) as SkillsIndexDict;
}

export function skillsIndexFromDecomposedDir(dir: string): SkillsIndexDict {
  return skillsIndexFromDecomposedDirNative(dir) as SkillsIndexDict;
}

export function mdToTree(
  markdownContent: string,
  sourcePath: string,
  config?: PageIndexConfigInput,
): Record<string, unknown> {
  return mdToTreeNative(
    markdownContent,
    sourcePath,
    resolveNativeConfig(config),
  ) as Record<string, unknown>;
}

export function getSkillDocument(
  documents: Record<string, unknown>,
  docId: string,
): Record<string, unknown> {
  return getSkillDocumentNative(documents, docId) as Record<string, unknown>;
}

export function getSkillStructure(
  documents: Record<string, unknown>,
  docId: string,
): unknown {
  return getSkillStructureNative(documents, docId);
}

export function getSkillLineContentFromSpec(
  index: SkillsIndexDict,
  docId: string,
  lineNumSpec: string,
): SkillLineContentRow[] {
  return getSkillLineContentFromSpecNative(
    index,
    docId,
    lineNumSpec,
  ) as SkillLineContentRow[];
}

function toNativeReconstructOptions(
  options?: ReconstructOptions,
): ReconstructOptionsNapi | undefined {
  if (!options) return undefined;
  return { keepAllHeaders: options.keepAllHeaders ?? false };
}

export function getSkillLineContent(
  index: SkillsIndexDict,
  docId: string,
  opts?: {
    lineNumSpecs?: string[];
    nodeIdSpecs?: string[];
  },
): SkillLineContentRow[] {
  return getSkillLineContentNative(
    index,
    docId,
    opts?.lineNumSpecs,
    opts?.nodeIdSpecs,
  ) as SkillLineContentRow[];
}

export function getSkillContentRetrieveResult(
  index: SkillsIndexDict,
  docId: string,
  opts?: {
    lineNumSpecs?: string[];
    nodeIdSpecs?: string[];
    options?: ReconstructOptions;
  },
): Record<string, unknown> {
  return getSkillContentRetrieveResultNative(
    index,
    docId,
    opts?.lineNumSpecs,
    opts?.nodeIdSpecs,
    toNativeReconstructOptions(opts?.options),
  ) as Record<string, unknown>;
}

export function reconstructSkillMarkdown(
  index: SkillsIndexDict,
  docId: string,
  opts?: {
    lineNumSpecs?: string[];
    nodeIdSpecs?: string[];
    options?: ReconstructOptions;
  },
): {
  markdown: string;
  matched_node_ids: number[];
  node_ids: number[];
  output_rel_path: string;
} {
  return reconstructSkillMarkdownNative(
    index,
    docId,
    opts?.lineNumSpecs,
    opts?.nodeIdSpecs,
    toNativeReconstructOptions(opts?.options),
  ) as {
    markdown: string;
    matched_node_ids: number[];
    node_ids: number[];
    output_rel_path: string;
  };
}

export function writeReconstructedSkill(
  catalogDir: string,
  index: SkillsIndexDict,
  docId: string,
  opts?: {
    lineNumSpecs?: string[];
    nodeIdSpecs?: string[];
    options?: ReconstructOptions;
  },
): string {
  return writeReconstructedSkillNative(
    catalogDir,
    index,
    docId,
    opts?.lineNumSpecs,
    opts?.nodeIdSpecs,
    toNativeReconstructOptions(opts?.options),
  );
}

export function parseSkillNodeIds(spec: string): number[] {
  return parseSkillNodeIdsNative(spec);
}

export function tokenCountFromDecomposedFrontmatter(
  content: string,
): number | null {
  return tokenCountFromDecomposedFrontmatterNative(content);
}

export function repairSkillNodes(
  entryDir: string,
  docId: string,
  config?: PageIndexConfigInput,
): void {
  repairSkillNodesNative(entryDir, docId, resolveNativeConfig(config));
}

export function buildPageIndexOnly(
  skillDirs: string[],
  config?: PageIndexConfigInput,
): SkillsIndexDict {
  return buildPageIndexOnlyNative(
    skillDirs,
    resolveNativeConfig(config),
  ) as SkillsIndexDict;
}

export function buildPageIndexForFile(
  sourcePath: string,
  config?: PageIndexConfigInput,
): SkillsIndexDict {
  return buildPageIndexForFileNative(
    sourcePath,
    resolveNativeConfig(config),
  ) as SkillsIndexDict;
}

export function pageIndexValid(
  entryDir: string,
  contentSha256: string,
): boolean {
  return pageIndexValidNative(entryDir, contentSha256);
}

export function loadSkillsIndexFromEntry(
  entryDir: string,
  docId: string,
): SkillsIndexDict {
  return loadSkillsIndexFromEntryNative(entryDir, docId) as SkillsIndexDict;
}

export function loadMergedSkillDocumentJson(
  entryDir: string,
  docId: string,
): Record<string, unknown> {
  return loadMergedSkillDocumentJsonNative(entryDir, docId) as Record<
    string,
    unknown
  >;
}

export function finalizeSkillDocumentJson(
  entryDir: string,
  docId: string,
  options: {
    pipeline: string;
    indexParams: Record<string, unknown>;
    sourcePath: string;
  },
): Record<string, unknown> {
  return finalizeSkillDocumentJsonNative(
    entryDir,
    docId,
    options.pipeline,
    options.indexParams,
    options.sourcePath,
  ) as Record<string, unknown>;
}

export function updateSkillDocumentSourcePath(
  entryDir: string,
  docId: string,
  sourcePath: string,
): Record<string, unknown> {
  return updateSkillDocumentSourcePathNative(
    entryDir,
    docId,
    sourcePath,
  ) as Record<string, unknown>;
}

export class SkillsBuilder {
  private inner: InstanceType<typeof SkillsBuilderNative>;

  constructor(options?: { memoryOnly?: boolean; outputDir?: string }) {
    this.inner = new SkillsBuilderNative(
      options?.memoryOnly ?? true,
      options?.outputDir,
    );
  }

  buildFromDirs(
    skillDirs: string[],
    config?: PageIndexConfigInput,
  ): SkillsIndexDict {
    return this.inner.buildFromDirs(
      skillDirs,
      resolveNativeConfig(config),
    ) as SkillsIndexDict;
  }

  buildFromFile(
    sourcePath: string,
    config?: PageIndexConfigInput,
  ): SkillsIndexDict {
    return this.inner.buildFromFile(
      sourcePath,
      resolveNativeConfig(config),
    ) as SkillsIndexDict;
  }

  writeCatalog(): SkillsIndexDict {
    return this.inner.writeCatalog() as SkillsIndexDict;
  }

  toSkillsIndexJson(): Record<string, unknown> {
    return this.inner.toSkillsIndexJson() as Record<string, unknown>;
  }

  toSkillsDict(): Record<string, unknown> {
    return this.inner.toSkillsDict() as Record<string, unknown>;
  }
}

File: sdk/python/src/chunk_your_skills/pageindex.py
"""Skills pageindex (markdown tree indexing and retrieval)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from chunk_your_skills._native import ReconstructOptions as _ReconstructOptions
from chunk_your_skills._native import SkillsBuilder as _SkillsBuilder
from chunk_your_skills._native import build_page_index_for_file as _build_page_index_for_file
from chunk_your_skills._native import build_page_index_only as _build_page_index_only
from chunk_your_skills._native import build_skills_index as _build_skills_index
from chunk_your_skills._native import finalize_skill_document_json as _finalize_skill_document_json
from chunk_your_skills._native import (
    get_skill_content_retrieve_result as _get_skill_content_retrieve_result,
)
from chunk_your_skills._native import get_skill_document as _get_skill_document
from chunk_your_skills._native import get_skill_line_content as _get_skill_line_content
from chunk_your_skills._native import (
    get_skill_line_content_from_spec as _get_skill_line_content_from_spec,
)
from chunk_your_skills._native import get_skill_structure as _get_skill_structure
from chunk_your_skills._native import load_merged_skill_document_json as _load_merged_skill_document_json
from chunk_your_skills._native import load_skills_index_from_dir as _load_skills_index_from_dir
from chunk_your_skills._native import load_skills_index_from_entry as _load_skills_index_from_entry
from chunk_your_skills._native import md_to_tree as _md_to_tree
from chunk_your_skills._native import page_index_valid as _page_index_valid
from chunk_your_skills._native import parse_skill_node_ids as _parse_skill_node_ids
from chunk_your_skills._native import reconstruct_skill_markdown as _reconstruct_skill_markdown
from chunk_your_skills._native import repair_skill_nodes as _repair_skill_nodes
from chunk_your_skills._native import (
    skills_index_from_decomposed_dir as _skills_index_from_decomposed_dir,
)
from chunk_your_skills._native import (
    token_count_from_decomposed_frontmatter as _token_count_from_decomposed_frontmatter,
)
from chunk_your_skills._native import parse_frontmatter_fields as _parse_frontmatter_fields
from chunk_your_skills._native import frontmatter_field as _frontmatter_field
from chunk_your_skills._native import (
    update_skill_document_source_path as _update_skill_document_source_path,
)
from chunk_your_skills._native import write_reconstructed_skill as _write_reconstructed_skill
from chunk_your_skills._native import write_skills_index as _write_skills_index


@dataclass
class ReconstructOptions:
    keep_all_headers: bool = False

    def to_native(self) -> _ReconstructOptions:
        return _ReconstructOptions(keep_all_headers=self.keep_all_headers)


@dataclass
class PageIndexConfig:
    if_add_node_id: bool = True
    if_add_node_text: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "if_add_node_id": self.if_add_node_id,
            "if_add_node_text": self.if_add_node_text,
        }

    @classmethod
    def from_mapping(cls, mapping: dict[str, Any] | None) -> PageIndexConfig:
        if not mapping:
            return default_page_index_config()
        cfg = default_page_index_config()
        if "if_add_node_id" in mapping:
            cfg.if_add_node_id = bool(mapping["if_add_node_id"])
        if "if_add_node_text" in mapping:
            cfg.if_add_node_text = bool(mapping["if_add_node_text"])
        return cfg


PageIndexConfigInput = PageIndexConfig | dict[str, Any]


def default_page_index_config() -> PageIndexConfig:
    return PageIndexConfig()


def page_index_config_from_mapping(mapping: dict[str, Any] | None = None) -> dict[str, Any]:
    """Partial pageindex settings from app/YAML; Rust merges unset keys with SDK defaults."""
    return PageIndexConfig.from_mapping(mapping).to_dict()


def build_skills_index(
    skill_dirs: list[str],
    *,
    config: PageIndexConfigInput | None = None,
) -> dict[str, Any]:
    cfg = _config_dict(config)
    return _build_skills_index(skill_dirs, cfg)


def write_skills_index(index: dict[str, Any], output_dir: str) -> None:
    _write_skills_index(index, output_dir)


def load_skills_index_from_dir(catalog_dir: str) -> dict[str, Any]:
    return _load_skills_index_from_dir(catalog_dir)


def skills_index_from_decomposed_dir(dir_path: str) -> dict[str, Any]:
    return _skills_index_from_decomposed_dir(dir_path)


def repair_skill_nodes(
    entry_dir: str,
    doc_id: str,
    *,
    config: PageIndexConfigInput | None = None,
) -> None:
    cfg = _config_dict(config)
    _repair_skill_nodes(entry_dir, doc_id, cfg)


def load_merged_skill_document_json(entry_dir: str, doc_id: str) -> dict[str, Any]:
    return _load_merged_skill_document_json(entry_dir, doc_id)


def build_page_index_for_file(
    source_path: str,
    *,
    config: PageIndexConfigInput | None = None,
) -> dict[str, Any]:
    cfg = _config_dict(config)
    return _build_page_index_for_file(source_path, cfg)


def build_page_index_only(
    skill_dirs: list[str],
    *,
    config: PageIndexConfigInput | None = None,
) -> dict[str, Any]:
    cfg = _config_dict(config)
    return _build_page_index_only(skill_dirs, cfg)


def page_index_valid(entry_dir: str, content_sha256: str) -> bool:
    return _page_index_valid(entry_dir, content_sha256)


def load_skills_index_from_entry(entry_dir: str, doc_id: str) -> dict[str, Any]:
    return _load_skills_index_from_entry(entry_dir, doc_id)


def finalize_skill_document_json(
    entry_dir: str,
    doc_id: str,
    *,
    pipeline: str,
    index_params: dict[str, Any],
    source_path: str,
) -> dict[str, Any]:
    return _finalize_skill_document_json(
        entry_dir,
        doc_id,
        pipeline=pipeline,
        index_params=index_params,
        source_path=source_path,
    )


def update_skill_document_source_path(
    entry_dir: str,
    doc_id: str,
    source_path: str,
) -> dict[str, Any]:
    return _update_skill_document_source_path(entry_dir, doc_id, source_path)


def md_to_tree(
    markdown_content: str,
    source_path: str,
    *,
    config: PageIndexConfigInput | None = None,
) -> dict[str, Any]:
    cfg = _config_dict(config)
    return _md_to_tree(markdown_content, source_path, cfg)


def get_skill_document(documents: dict[str, Any], doc_id: str) -> dict[str, Any]:
    return _get_skill_document(documents, doc_id)


def get_skill_structure(documents: dict[str, Any], doc_id: str) -> list[Any] | dict[str, Any]:
    return _get_skill_structure(documents, doc_id)


def get_skill_line_content_from_spec(
    index: dict[str, Any],
    doc_id: str,
    line_num_spec: str,
) -> list[dict[str, Any]]:
    return _get_skill_line_content_from_spec(index, doc_id, line_num_spec)


def get_skill_line_content(
    index: dict[str, Any],
    doc_id: str,
    *,
    line_num_specs: list[str] | None = None,
    node_id_specs: list[str] | None = None,
) -> list[dict[str, Any]]:
    return _get_skill_line_content(
        index,
        doc_id,
        line_num_specs=line_num_specs,
        node_id_specs=node_id_specs,
    )


def get_skill_content_retrieve_result(
    index: dict[str, Any],
    doc_id: str,
    *,
    line_num_specs: list[str] | None = None,
    node_id_specs: list[str] | None = None,
    options: ReconstructOptions | dict[str, Any] | None = None,
) -> dict[str, Any]:
    return _get_skill_content_retrieve_result(
        index,
        doc_id,
        line_num_specs=line_num_specs,
        node_id_specs=node_id_specs,
        options=_reconstruct_options_native(options),
    )


def reconstruct_skill_markdown(
    index: dict[str, Any],
    doc_id: str,
    *,
    line_num_specs: list[str] | None = None,
    node_id_specs: list[str] | None = None,
    options: ReconstructOptions | dict[str, Any] | None = None,
) -> dict[str, Any]:
    return _reconstruct_skill_markdown(
        index,
        doc_id,
        line_num_specs=line_num_specs,
        node_id_specs=node_id_specs,
        options=_reconstruct_options_native(options),
    )


def write_reconstructed_skill(
    catalog_dir: str,
    index: dict[str, Any],
    doc_id: str,
    *,
    line_num_specs: list[str] | None = None,
    node_id_specs: list[str] | None = None,
    options: ReconstructOptions | dict[str, Any] | None = None,
) -> str:
    return _write_reconstructed_skill(
        catalog_dir,
        index,
        doc_id,
        line_num_specs=line_num_specs,
        node_id_specs=node_id_specs,
        options=_reconstruct_options_native(options),
    )


def parse_skill_node_ids(spec: str) -> list[int]:
    return _parse_skill_node_ids(spec)


def token_count_from_decomposed_frontmatter(content: str) -> int | None:
    """Parse ``token_count`` from decomposed markdown/JSON frontmatter when present."""
    return _token_count_from_decomposed_frontmatter(content)


def parse_frontmatter_fields(content: str) -> list[dict[str, Any]] | None:
    """Parse root-level YAML frontmatter keys into semantic values."""
    return _parse_frontmatter_fields(content)


def frontmatter_field(content: str, key: str) -> Any:
    """Look up one semantically parsed frontmatter field by name."""
    return _frontmatter_field(content, key)


class SkillsBuilder:
    def __init__(self, *, memory_only: bool = True, output_dir: str | None = None) -> None:
        self._inner = _SkillsBuilder(memory_only=memory_only, output_dir=output_dir)

    def build_from_dirs(
        self,
        skill_dirs: list[str],
        *,
        config: PageIndexConfigInput | None = None,
    ) -> dict[str, Any]:
        cfg = _config_dict(config)
        return self._inner.build_from_dirs(skill_dirs, cfg)

    def build_from_file(
        self,
        source_path: str,
        *,
        config: PageIndexConfigInput | None = None,
    ) -> dict[str, Any]:
        cfg = _config_dict(config)
        return self._inner.build_from_file(source_path, cfg)

    def write_catalog(self) -> dict[str, Any]:
        return self._inner.write_catalog()

    def to_skills_index_json(self) -> dict[str, Any]:
        return self._inner.to_skills_index_json()

    def to_skills_dict(self) -> dict[str, Any]:
        return self._inner.to_skills_dict()


def _config_dict(config: PageIndexConfigInput | None) -> dict[str, Any] | None:
    if config is None:
        return None
    if isinstance(config, PageIndexConfig):
        return config.to_dict()
    if isinstance(config, dict) and _is_snake_case_pageindex_dict(config):
        return config
    if isinstance(config, dict):
        return PageIndexConfig.from_mapping(config).to_dict()
    return config


def _is_snake_case_pageindex_dict(config: dict[str, Any]) -> bool:
    return any(key in config for key in ("if_add_node_id", "if_add_node_text"))


def _reconstruct_options_native(
    options: ReconstructOptions | dict[str, Any] | None,
) -> _ReconstructOptions | None:
    if options is None:
        return None
    if isinstance(options, ReconstructOptions):
        return options.to_native()
    return _ReconstructOptions(
        keep_all_headers=bool(options.get("keep_all_headers", options.get("keepAllHeaders", False))),
    )
